src/main/java/sparksoniq/jsoniq/runtime/iterator/operational/TypePromotionIterator.java

```diff
@@ -0,0 +1,106 @@
+package sparksoniq.jsoniq.runtime.iterator.operational;
+
+import org.apache.spark.api.java.JavaRDD;
+import org.apache.spark.api.java.function.Function;
+import org.rumbledb.api.Item;
+import sparksoniq.exceptions.TreatException;
+import sparksoniq.exceptions.UnexpectedTypeException;
+import sparksoniq.jsoniq.item.AtomicItem;
+import sparksoniq.jsoniq.runtime.iterator.HybridRuntimeIterator;
+import sparksoniq.jsoniq.runtime.iterator.RuntimeIterator;
+import sparksoniq.jsoniq.runtime.iterator.functions.sequences.general.TypePromotionClosure;
+import sparksoniq.jsoniq.runtime.metadata.IteratorMetadata;
+import sparksoniq.semantics.DynamicContext;
+import sparksoniq.semantics.types.SequenceType;
+
+import java.util.Collections;
+
+public class TypePromotionIterator extends HybridRuntimeIterator {
+
+    private static final long serialVersionUID = 1L;
+    private final String _exceptionMessage;
+    private SequenceType _sequenceType;
+    private TreatIterator _treatIterator;
+    private RuntimeIterator _child;
+
+    public TypePromotionIterator(
+            String exceptionMessage,
+            RuntimeIterator iterator,
+            SequenceType sequenceType,
+            IteratorMetadata iteratorMetadata
+    ) {
+        super(Collections.singletonList(iterator), iteratorMetadata);
+        this._exceptionMessage = exceptionMessage;
+        this._child = iterator;
+        this._sequenceType = sequenceType;
+        this._treatIterator = new TreatIterator(_child, _sequenceType, true, iteratorMetadata);
+    }
+
+    @Override
+    public void openLocal() {
+        try {
+            _treatIterator.open(_currentDynamicContext);
+        } catch (TreatException e) {
+            this.checkTypePromotion(e.getJSONiqErrorMessage());
+        }
+    }
+
+    @Override
+    public void closeLocal() {
+        _treatIterator.close();
+    }
+
+    @Override
+    public void resetLocal(DynamicContext context) {
+        _treatIterator.reset(context);
+    }
+
+    @Override
+    public boolean hasNextLocal() {
+        return _treatIterator.hasNext();
+    }
+
+    @Override
+    public Item nextLocal() {
+        Item result;
+        try {
+            result = _treatIterator.next();
+        } catch (TreatException e) {
+            this.checkTypePromotion(e.getJSONiqErrorMessage());
+            result = _treatIterator._currentResult;
```

Comment:
I see two small but important issues here:
1. Breach of encapsulation in the iterator API (the state of an iterator should not be modified by the consumer).
2. Use of an exception in "normal use" of the engine (that is, without the query having any issue).

In order to solve them, there is a trick commonly used by other iterator-based XQuery/JSONiq engines (so we do not need to reinvent a new design): swapping the treat as and promotion iterators, i.e., having the treat-as iterator calling the promotion iterator rather than the other way round.

That way,
- the promotion iterator's only job is to check whether each item that doesn't match the target type can be promoted or not (with the new canBePromotedTo method you introduced), and if so, promote it.
- the treat as iterator's only job is to check for types (as it already does) and always returns sequences that match the target type (no breach of contract).

With this clean separation of concerns between promotion and type checking, the iterators can be chained completely from outside, in the function catalogue, with no need from one (promotion) to internally instantiate the other (treat as). See my other comments in the course catalogue.

Then
The TreatAsException will only be thrown if no promotion happened in the child iterator, and does not need to be caught, which solves issue 2.
And the TypePromotionIterator no longer needs to alter the state of the TypePromotionIterator, as the latter can take care of deciding to promote all by itself, which solves issue 1.

Reply:
If I understand correctly, in order to do what you suggest, I would have to basically copy the same behavior of TreatIterator in TypePromotionIterator and add the check for promotion there. As a matter of fact, before checking for promotion I would have to perform the check on the type (_isTypeOf_) that TreatIterator is doing, and I would also have to check for correct arity (and check on empty sequence) i.e. all the checks that TreatIterator does now. The only difference between the two would be that TypePromotionIterator checks for promotion in case the check _isTypeOf_ is false and promote if possible. 
If that is what we end up doing, though, we do not actually need to encapsulate such type of iterator with TreatIterator again.

Reply:
> If that is what we end up doing, though, we do not actually need to encapsulate such type of iterator with TreatIterator again.

Correct -- I would suggest to give it a try. That's good thinking.

One way to do that could be to either clone the treat as iterator and extend the cloned iterator with promotion if a type check fails. Then, only use this cloned iterator.

Another way is to add a boolean parameter to the treat as iterator and extend it with promotion if that boolean is true., Then, use this treat as iterator setting the boolean to true.

In both cases, only a single iterator would suffice.

Reply:
In order to keep to keep the two iterator separate (as they are also throwing exceptions with a different message), I went for the first option. Even thought there is some code duplication, I think this disentanglement makes the iterator hierarchy cleaner and we get the separation of concerns we aimed for. 
As a side benefit, we also got rid of one layer of additional and unnecessary iterator.